### PR TITLE
fix(core-p2p): parallel download

### PR DESCRIPTION
**Pull request by vasild**
## Checklist

- [ ] Documentation _(if necessary)_
- [ ] Tests _(if necessary)_
- [ ] Ready to be merged

### CODE REVIEW

File: packages/core-p2p/src/network-monitor.ts [hunk 1]
```diff
@@ -301,57 +300,106 @@ export class NetworkMonitor implements P2P.INetworkMonitor {
...
+        for (let i = 0; i < chunksToDownload; i++) {
+            const height: number = fromBlockHeight + chunkSize * i;
+            const isLastChunk: boolean = i === chunksToDownload - 1;
+            const blocksRange: string = `[${height + 1}, ${isLastChunk ? ".." : height + 1 + chunkSize}]`;
+
+            downloadJobs.push(async () => {
+                let blocks: Interfaces.IBlockData[];
+                let peer: P2P.IPeer;
+
+                for (peer of shuffle(peersNotForked)) {
```
Comment (air1one) [inline packages/core-p2p/src/network-monitor.ts]: For every chunk you start from a "shuffled" list of peers, so the 1st peer you download from can be the same for say 5 chunks : you will hit rate limit and be disconnected from the peer, causing the 5 downloads to fail (then yes I understand that you pick another peer but then you have more chances to hit rate limit again and cause some cascading download errors due to rate limit).
I'd say at least for the 1st try for every chunk to be downloaded from a different peer, then somehow be careful with rate limit for the "fallback" peers you download from if 1st one failed.

Comment (vasild) [reply packages/core-p2p/src/network-monitor.ts]: Right, good point! Should be addressed in 21e3d0036.

"Resolve conversation" this if you are ok with it.

Comment (air1one) [reply packages/core-p2p/src/network-monitor.ts]: yes looking good now :ok_hand:

File: packages/core-p2p/src/network-monitor.ts [hunk 2]
```diff
@@ -301,57 +300,113 @@ export class NetworkMonitor implements P2P.INetworkMonitor {
...
-                        }
+        this.logger.debug(`Downloading blocks in chunks: ${chunksHumanReadable}`);
 
-                        return blocks;
-                    }),
-            )).reduce((acc, curr) => [...acc, ...(curr || [])], []);
+        let firstFailureMessage: string;
+
+        try {
+            await Promise.all(downloadJobs.map(f => f()));
```
Comment (supaiku0) [inline packages/core-p2p/src/network-monitor.ts]: The `.map(f => f())`  can be removed, since `Promise.all` does that internally.

Comment (vasild) [reply packages/core-p2p/src/network-monitor.ts]: That is necessary to convert the async functions to promises by calling them.

**Comment by supaiku0** [comment]
There's one issue now, it tries really hard to get blocks.

This is after reset to height 4.000.000 and then syncing without `core-magistrate-transactions` loaded, so it gets stuck at one point because it refuses the new tx type:
```
2019-10-23 00:44:32.200] DEBUG: event 'NOTSYNCED': {"syncWithNetwork":"syncing"} -> {"syncWithNetwork":"downloadBlocks"} -> actions: [downloadBlocks]
[2019-10-23 00:44:32.202] DEBUG: Downloading blocks in chunks: [4005942, 4006341], [4006342, 4006741], [4006742, 4007141], [4007142, 4007541], [4007542, 4007941], [4007942, 4008341], [4008342, 4008741], [4008742, 4009141], [4009142, 4009541], [4009542, ..]
[2019-10-23 00:44:32.373] DEBUG: Downloaded blocks [4008342, 4008741] (400) from 167.114.225.21:4002
[2019-10-23 00:44:32.384] DEBUG: Downloaded blocks [4007142, 4007541] (400) from 167.114.224.71:4002
[2019-10-23 00:44:32.390] DEBUG: Downloaded blocks [4009542, ..] (400) from 149.202.183.157:4002
[2019-10-23 00:44:32.468] DEBUG: Downloaded blocks [4009142, 4009541] (400) from 54.38.138.244:4002
[2019-10-23 00:44:32.479] DEBUG: Downloaded blocks [4008742, 4009141] (400) from 51.75.168.217:4002
[2019-10-23 00:44:32.488] DEBUG: Failed to download blocks from height 4,006,741 via 167.114.230.48.
[2019-10-23 00:44:32.662] DEBUG: Failed to download blocks from height 4,006,741 via 149.202.181.70.
[2019-10-23 00:44:32.666] DEBUG: Failed to download blocks from height 4,007,941 via 198.46.160.94.
[2019-10-23 00:44:32.788] DEBUG: Failed to download blocks from height 4,007,941 via 167.114.254.197.
[2019-10-23 00:44:32.822] DEBUG: Failed to download blocks from height 4,006,341 via 167.114.43.45.
[2019-10-23 00:44:32.991] DEBUG: Failed to download blocks from height 4,006,341 via 137.74.27.246.
[2019-10-23 00:44:32.996] DEBUG: Downloaded blocks [4007542, 4007941] (400) from 139.99.79.180:4002
[2019-10-23 00:44:33.157] DEBUG: Failed to download blocks from height 4,006,341 via 54.36.116.38.
[2019-10-23 00:44:33.189] DEBUG: Failed to download blocks from height 4,006,741 via 138.197.79.32.
[2019-10-23 00:44:33.376] DEBUG: Failed to download blocks from height 4,007,941 via 139.99.78.153.
[2019-10-23 00:44:33.389] DEBUG: Failed to download blocks from height 4,006,741 via 54.36.116.164.
[2019-10-23 00:44:33.395] DEBUG: Failed to download blocks from height 4,006,341 via 145.239.29.252.
[2019-10-23 00:44:33.567] DEBUG: Failed to download blocks from height 4,006,741 via 149.202.160.132.
[2019-10-23 00:44:33.582] DEBUG: Failed to download blocks from height 4,006,341 via 167.99.92.224.
[2019-10-23 00:44:33.730] DEBUG: Failed to download blocks from height 4,006,741 via 
...
[2019-10-23 00:45:18.028] ERROR: Could not download blocks [4007942, 4008341] from any of 126 peer(s). Last attempt returned 0 block(s) from peer 139.99.218.175.
[2019-10-23 00:45:18.029] INFO : Could not download any blocks from any peer from height 4005942
[2019-10-23 00:45:18.036] DEBUG: event 'NOBLOCK': {"syncWithNetwork":"downloadBlocks"} -> {"syncWithNetwork":"syncing"} -> actions: [checkLastDownloadedBlockSynced]
[2019-10-23 00:45:18.036] DEBUG: Queued chunks of blocks (process: 0)
[2019-10-23 00:45:18.037] DEBUG: event 'NOTSYNCED': {"syncWithNetwork":"syncing"} -> {"syncWithNetwork":"downloadBlocks"} -> actions: [downloadBlocks]
[2019-10-23 00:45:18.039] DEBUG: Downloading blocks in chunks: [4005942, 4006341], [4006342, 4006741], [4006742, 4007141], [4007142, 4007541], [4007542, 4007941], [4007942, 4008341], [4008342, 4008741], [4008742, 4009141], [4009142, 4009541], [4009542, ..]
[2019-10-23 00:45:18.230] DEBUG: Failed to download blocks from height 4,006,341 via 167.114.43.35.
...

It eventually stops for a moment, but look at the timestamp how long it took.
```

**Pull request merged**